### PR TITLE
Reduce number of read calls when not asking for progress reporting.

### PR DESCRIPTION
**Pull request by parkera**
Also, when asking for progress reporting, cap the number of updates at approximately 100 instead of fileSize/4k.

### CODE REVIEW

File: Sources/FoundationEssentials/Data/Data+Reading.swift [hunk 1]
```diff
@@ -370,7 +369,7 @@ internal func readBytesFromFile(path inPath: PathOrURL, reportProgress: Bool, ma
         
         localProgress?.becomeCurrent(withPendingUnitCount: Int64(fileSize))
         do {
-            let length = try readBytesFromFileDescriptor(fd, path: inPath, buffer: bytes, length: fileSize, chunkSize: size_t(preferredChunkSize), reportProgress: reportProgress)
```
Comment (parkera) [inline Sources/FoundationEssentials/Data/Data+Reading.swift:373]: The block size is just too small (4k on macOS, e.g.), and results in a very large number of `read` calls.

**Review by kperryua — APPROVED** [review]

**Comment by jmschonfeld** [comment]
The linux test failure looks related:

```
Test Case 'DataIOTests.test_largeFile' started at 2024-05-20 18:29:52.910
/build/swift-foundation/Tests/FoundationEssentialsTests/DataIOTests.swift:215: error: DataIOTests.test_largeFile : XCTAssertEqual failed: ("2147549184") is not equal to ("2147479552") - 
Test Case 'DataIOTests.test_largeFile' failed (17.749 seconds)
```

**Comment by parkera** [comment]
@swift-ci test

**Comment by parkera** [comment]
Updated to provide a "read until EOF" option, true by default. On Darwin we were always reading the requested `0x7ffffff`, but on Linux we were reading `0x7fff000`, then failing to loop again. This mirrors the implementation from swift-corelibs-foundation.

**Comment by itingliu** [comment]
LGTM! Can we merge this?

**Comment by parkera** [comment]
Yah, I wanted @kperryua to take a second look if possible.

**Comment by kperryua** [comment]
> Yah, I wanted @kperryua to take a second look if possible.

Took me a bit to grok what the readUntilEOF change is about. I take a little issue with the name, since it presumes that the `length` parameter is expected to be EOF of a regular file. That's only technically true in this particular case of Data reading, since we're first querying the length of the file on disk and then passing that parameter for length. In theory one could call this function with a length less than the full length of the file, if one has a limited buffer.

**Comment by kperryua** [comment]
> it's more like `readUntilLength`

I like that better.

**Review by kperryua — APPROVED** [review]

**Pull request merged**